### ISSUE TITLE
Is it possible to inspect training progress with Docker?

### ISSUE THREAD

**Issue by zhenyulin** [NVIDIA/nvidia-docker#320]
I'm currently using docker/nvidia-docker to run Deep Learning stacks.

I'm currently running the training with `docker exec -it <container_name> bash` to enter the container and use `Ctrl+P` then `Ctrl+Q` to escape from the container after starting the training script.

Using `docker logs <container_name>` will only show the logs when the training is finished, whereas I haven't been able to track the training progress, so that I can planning other tasks on the machine.

Was wondering if there's way to inspect training progress with docker?

**Comment by KleinYuan** [comment]
@zhenyulin `docker logs -f <container_name>` => https://docs.docker.com/engine/reference/commandline/logs/#/options

**Comment by zhenyulin** [comment]
@KleinYuan yes, this works for me in the case of following the logs printout by a web server, but doesn't work to follow the logs in the training container. Is it caused by the way how the logs are implemented in the training script?

**Comment by flx42** [comment]
@zhenyulin you should run your training when starting the container, not by doing an `docker exec` afterwards, it will solve your problem.
For instance:
```
nvidia-docker run -d nvidia/caffe caffe train [...]
```

**Comment by zhenyulin** [comment]
@flx42 thanks, this works! 

In this case, I will be creating new container every time I run training, while I'm currently storing training data and output model in the container, and potentially modify the script a bit on the go. 

Would it be possible to reuse a container but still be able to get the logs?

**Comment by flx42** [comment]
> In this case, I will be creating new container every time I run training, while I'm currently storing training data and output model in the container, and potentially modify the script a bit on the go.

I don't recommend using containers that way. Instead, you should run your container as read-only, and all input/output operations should be to/from volumes mounted from this host. You should use the container filesystem to persist your training output, and it's actually detrimental for performance to do so (because the I/O operations will be on the union file system, and that's slower).

**Comment by flx42** [comment]
Or, just run your multiple trainings from your initial docker run:
```
$ nvidia-docker run -ti nvidia/caffe bash
# caffe train model1
[...]
# caffe train model2
[...]
```
No need to `docker exec`.

**Comment by zhenyulin** [comment]
@flx42 thanks for help out and this approach looks more reasonable than my previous ones. Without `docker exec` , it seems difficult if I want to tune the training code in the container during the process, e.g. encoder. Do you have a suggestion on how to tackle this problem?

**Comment by zhenyulin** [comment]
In this approach, it would be nice to have autocompletion of image names over `nvidia-docker run`, as I found `docker exec <container_name>` autocompletion provides a very good experience during the tweaking. 😌

**Comment by flx42** [comment]
@zhenyulin in this case, use `docker exec` to edit your code, but use your main bash to run the code.
What you are doing is a little weird, you shouldn't try to use `docker logs` AND `docker exec`

To get auto-completion with `nvidia-docker`, copy the file `/usr/share/bash-completion/completions/docker` to `/usr/share/bash-completion/completions/nvidia-docker` and modify the last line of this new file, and that's it.

**Comment by zhenyulin** [comment]
@flx42 thanks for the advice! 

The tricky bits are:
- `docker exec` can only be applied to a <container> 
- if by main bash you mean using `nvidia-docker run --rm -ti <image> command`, the code edits will basically be remaining in the container rather than updated to the image, and I can't 'run' a container.

Maybe it is more of a docker question, but I was wondering what's the difference between `run` and `exec` towards logs?

Regarding the auto-completion, I've copied the `docker.bash-completion` file to `usr/local/etc/bash_completion.d` on Mac, shall I replace all `docker` command names with `nvidia-docker` when editing the file?

**Comment by dashesy** [comment]
Can pass `--log-driver=journald` and [use journalctl](https://docs.docker.com/engine/admin/logging/journald/#labels-and-env). 
Also, bind a directory with your code `--volume ~/code:/code`, then you can tune it outside the container. When you are comfortable you can move it to the image.
